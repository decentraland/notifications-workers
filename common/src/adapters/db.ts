import SQL, { SQLStatement } from 'sql-template-strings'
import { SubscriptionDB } from '../types'
import { IPgComponent } from '@well-known-components/pg-component'
import { defaultSubscription } from '../subscriptions'

export type DbComponents = {
  pg: IPgComponent
}

export type DbComponent = {
  findSubscription(address: string): Promise<SubscriptionDB>
}

export function createDbComponent({ pg }: Pick<DbComponents, 'pg'>): DbComponent {
  async function findSubscription(address: string): Promise<SubscriptionDB> {
    const query: SQLStatement = SQL`
        SELECT address,
               email,
               details,
               created_at,
               updated_at
        FROM subscriptions n
        WHERE address = ${address.toLowerCase()}
    `

    const result = await pg.query<SubscriptionDB>(query)
    if (result.rowCount === 0) {
      return {
        address: address.toLowerCase(),
        email: undefined,
        details: defaultSubscription(),
        created_at: Date.now(),
        updated_at: Date.now()
      }
    }
    return result.rows[0]
  }

  return {
    findSubscription
  }
}
