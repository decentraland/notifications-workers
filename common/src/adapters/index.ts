export * from './db'
export * from './sendgrid-client'
export * from './data-warehouse-client'
export * from './profiles'
