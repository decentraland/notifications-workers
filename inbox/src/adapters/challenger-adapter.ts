import { IBaseComponent } from '@well-known-components/interfaces'
import { AppComponents, Feature } from '../types'
import { EthAddress } from '@dcl/schemas'

export type IChallengerAdapter = IBaseComponent & {
  verifyChallengeIfEnabled: (token: string | undefined, userAddress: EthAddress) => Promise<boolean>
}

export async function createChallengerAdapter(
  components: Pick<AppComponents, 'logs' | 'config' | 'fetch' | 'featureFlagsAdapter'>
) {
  const { logs, config, fetch, featureFlagsAdapter } = components

  const CLOUDFLARE_CHALLENGE_URL = 'https://challenges.cloudflare.com/turnstile/v0/siteverify'
  const CLOUDFLARE_SECRET = await config.requireString('CLOUDFLARE_SECRET')
  const logger = logs.getLogger('challenger-adapter')

  /**
   * Checks if the challenge is enabled and if the token is correct
   * @param token - The token to verify
   * @param userAddress - The address of the user
   * @returns True if the challenge is disabled or the token is correct
   */
  async function verifyChallengeIfEnabled(token: string | undefined, userAddress: EthAddress) {
    const isChallengeEnabled = featureFlagsAdapter.isEnabled(Feature.TURNSTILE_VERIFICATION)

    logger.debug('CloudFlare challenge is enabled, verifying token', {
      userAddress,
      token: token ?? 'missing',
      isChallengeEnabled: isChallengeEnabled ? 'enabled' : 'disabled'
    })

    if (!isChallengeEnabled) {
      return true
    }

    if (!token) {
      return false
    }

    const response = await fetch.fetch(CLOUDFLARE_CHALLENGE_URL, {
      method: 'POST',
      body: JSON.stringify({
        secret: CLOUDFLARE_SECRET,
        response: token
      })
    })

    const data = await response.json()

    return data.success
  }

  return {
    verifyChallengeIfEnabled
  }
}
