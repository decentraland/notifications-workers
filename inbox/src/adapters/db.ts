import SQL, { SQLStatement } from 'sql-template-strings'
import { AppComponents, DbNotification, NotificationEvent } from '../types'

export type DbComponent = {
  findNotifications(users: string[], onlyUnread: boolean, from: number, limit: number): Promise<DbNotification[]>
  markNotificationsAsRead(userId: string, notificationIds: string[]): Promise<number>
}

export function createDbComponent({ pg }: Pick<AppComponents, 'pg' | 'logs'>): DbComponent {
  async function findNotifications(
    users: string[],
    onlyUnread: boolean,
    from: number,
    limit: number
  ): Promise<DbNotification[]> {
    const query: SQLStatement = SQL`
        SELECT id,
               event_key,
               type,
               n.address as address,
               metadata,
               timestamp,
               n.read_at as read_at,
               created_at,
               updated_at,
               br.address AS broadcast_address,
               br.read_at AS broadcast_read_at
        FROM notifications n
        LEFT JOIN broadcast_read br ON n.id = br.notification_id
    `

    const lowercaseUsers = users.map((u) => u.toLowerCase())
    const whereClause: SQLStatement[] = [SQL`(n.address IS NULL OR n.address = ANY (${lowercaseUsers}))`]
    if (from > 0) {
      whereClause.push(SQL`timestamp >= ${from}`)
    }
    if (onlyUnread) {
      whereClause.push(SQL`(n.address IS NOT NULL AND n.read_at IS NULL) OR (n.address IS NULL AND br.read_at IS NULL)`)
    }
    let where = SQL` WHERE `.append(whereClause[0])
    for (const condition of whereClause.slice(1)) {
      where = where.append(' AND ').append(condition)
    }

    query.append(where)
    query.append(SQL` ORDER BY timestamp DESC`)
    query.append(SQL` LIMIT ${limit}`)

    return (await pg.query<DbNotification>(query)).rows
  }

  async function markNotificationsAsRead(userId: string, notificationIds: string[]) {
    const readAt = Date.now()

    const updateResult = await pg.query<NotificationEvent>(SQL`
        UPDATE notifications
        SET    read_at    = ${readAt},
               updated_at = ${readAt}
        WHERE  read_at IS NULL
          AND  address = ${userId.toLowerCase()}
          AND  id = ANY (${notificationIds})
        RETURNING id
    `)
    let notificationCount = updateResult.rowCount

    const addressedNotificationsIds = new Set(updateResult.rows.map((n) => n.id))
    const potentialBroadcastIds = notificationIds.filter((id) => !addressedNotificationsIds.has(id))
    if (potentialBroadcastIds.length > 0) {
      const lowerUserId = userId.toLowerCase()
      const query = SQL`
        INSERT INTO broadcast_read (notification_id, address, read_at)
          SELECT id, ${lowerUserId}, ${readAt}
          FROM   notifications
          WHERE  id = ANY (${potentialBroadcastIds})
            AND  address IS NULL
        ON CONFLICT (notification_id, address) DO NOTHING`

      notificationCount += (await pg.query<NotificationEvent>(query)).rowCount
    }

    return notificationCount
  }

  return {
    findNotifications,
    markNotificationsAsRead
  }
}
