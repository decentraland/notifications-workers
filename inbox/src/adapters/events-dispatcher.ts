import { randomUUID } from 'node:crypto'
import { AppComponents } from '../types'
import { Readable } from 'node:stream'
import { NotificationEvent } from '@notifications/commons'

export type Session = {
  userId: string
  stream: Readable
}

export type EventsDispatcherComponent = {
  start(): void
  stop(): void
  addClient(s: Session): string
  removeClient(uuid: string): void
}

export function createEventsDispatcherComponent({
  logs,
  db
}: Pick<AppComponents, 'db' | 'logs'>): EventsDispatcherComponent {
  const logger = logs.getLogger('events-dispatcher')

  const clients = new Map<string, Session>()

  function addClient(s: Session): string {
    logger.debug('add client')
    const uuid = randomUUID()
    clients.set(uuid, s)
    return uuid
  }

  function removeClient(uuid: string): void {
    logger.debug('remove client')
    clients.delete(uuid)
  }

  let interval: any
  function start() {
    let lastCheck = 0
    interval = setInterval(async () => {
      const users = new Set(Array.from(clients.values()).map((s) => s.userId))
      const notifications = await db.findNotifications(Array.from(users), true, 100, lastCheck)

      const notificationsByUser = new Map<string, NotificationEvent[]>()

      for (const notification of notifications) {
        const userNotifications: NotificationEvent[] = notificationsByUser.get(notification.address) ?? []
        userNotifications.push(notification)
        notificationsByUser.set(notification.address, userNotifications)
      }

      for (const [address, notifications] of notificationsByUser) {
        for (const { userId, stream } of clients.values()) {
          if (userId === address) {
            stream.push(`event: ping\n\n`)

            // Events are ordered and retrieved by the local timestamp of the database/server, not from the notification origin

            for (const notification of notifications) {
              stream.push(`data: ${JSON.stringify(notification)}\n\n`)
            }
          }
        }
      }

      // Every interval, send a "ping" event.
      lastCheck = Date.now()
    }, 15000)
  }

  function stop() {
    if (interval) {
      clearInterval(interval)
    }
  }

  return {
    addClient,
    removeClient,
    start,
    stop
  }
}
