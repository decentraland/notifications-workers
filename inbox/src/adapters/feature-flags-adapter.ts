import { IBaseComponent, START_COMPONENT, STOP_COMPONENT } from '@well-known-components/interfaces'
import { AppComponents, Feature } from '../types'
import { ApplicationName } from '@well-known-components/features-component'

export type IFeatureFlagsAdapter = IBaseComponent & {
  isEnabled: (feature: Feature) => boolean
}

// How often to refresh feature flags (4 minutes in milliseconds)
const FEATURE_FLAG_REFRESH_INTERVAL = 4 * 60 * 1000

export async function createFeatureFlagsAdapter(components: Pick<AppComponents, 'logs' | 'features'>) {
  const { logs, features } = components

  const logger = logs.getLogger('feature-flags-adapter')
  const featuresFlagMap = new Map<Feature, boolean>()

  let refreshInterval: NodeJS.Timeout | null = null

  async function refreshFeatureFlags() {
    try {
      // Get Turnstile Verification feature flag
      const isTurnstileVerificationEnabled = await features.getIsFeatureEnabled(
        ApplicationName.DAPPS,
        Feature.TURNSTILE_VERIFICATION
      )
      logger.debug('Refreshed Turnstile Verification feature flag', {
        isEnabled: isTurnstileVerificationEnabled ? 'enabled' : 'disabled'
      })

      featuresFlagMap.set(Feature.TURNSTILE_VERIFICATION, isTurnstileVerificationEnabled)
    } catch (error) {
      logger.error('Failed to refresh feature flags', {
        error: error instanceof Error ? error.message : String(error)
      })
    }
  }

  /**
   * Start component and initialize periodic feature flag refresh
   */
  async function start() {
    logger.info('Starting feature flags adapter')

    // Do initial refresh
    await refreshFeatureFlags()

    // Set up periodic refresh
    refreshInterval = setInterval(async () => {
      await refreshFeatureFlags()
    }, FEATURE_FLAG_REFRESH_INTERVAL)

    logger.info('Feature flags adapter started', {
      refreshInterval: FEATURE_FLAG_REFRESH_INTERVAL / 1000 / 60 + ' minutes'
    })
  }

  /**
   * Stop component and clear interval
   */
  async function stop() {
    logger.info('Stopping feature flags adapter')

    if (refreshInterval) {
      clearInterval(refreshInterval)
      refreshInterval = null
    }
  }

  return {
    [START_COMPONENT]: start,
    [STOP_COMPONENT]: stop,
    refreshFeatureFlags,
    isEnabled: (feature: Feature) => {
      return featuresFlagMap.get(feature) ?? false
    }
  }
}
