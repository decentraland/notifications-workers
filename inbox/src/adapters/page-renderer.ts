import * as fs from 'node:fs'
import * as path from 'path'
import handlebars from 'handlebars'

export type IPageRenderer = {
  renderPage(template: string, context: any): string
}

function loadTemplates() {
  return fs
    .readdirSync(path.join(__dirname, 'page-templates'), { withFileTypes: true })
    .filter((file) => file.isFile() && file.name.endsWith('.handlebars'))
    .reduce(
      (acc, file) => {
        console.log('file', file, file.parentPath, file.name)
        acc[file.name.replace('.handlebars', '')] = handlebars.compile(
          fs.readFileSync(path.join(file.parentPath, file.name), 'utf8')
        )
        return acc
      },
      {} as Record<string, HandlebarsTemplateDelegate>
    )
}

export async function createPageRenderer(): Promise<IPageRenderer> {
  const templates = loadTemplates()

  function renderPage(template: string, context: any): string {
    return templates[template](context)
  }

  return {
    renderPage
  }
}
