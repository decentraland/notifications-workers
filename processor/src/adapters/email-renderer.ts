import * as fs from 'node:fs'
import * as path from 'path'
import handlebars from 'handlebars'
import { NotificationType } from '@dcl/schemas'
import { Email } from '@notifications/common'
import { AppComponents, IEmailRenderer, NotificationRecord } from '../types'
import { formatMana } from '../logic/utils'

const nonTransformer = (notification: NotificationRecord): NotificationRecord => notification

const formatPriceAsMana = (notification: NotificationRecord) => ({
  ...notification,
  metadata: {
    ...notification.metadata,
    price: formatMana(notification.metadata.price)
  }
})

const formatRoyaltiesCutAsMana = (notification: NotificationRecord) => ({
  ...notification,
  metadata: {
    ...notification.metadata,
    royaltiesCut: formatMana(notification.metadata.royaltiesCut)
  }
})

const transformers: Partial<Record<NotificationType, (notification: NotificationRecord) => NotificationRecord>> = {
  [NotificationType.BID_ACCEPTED]: formatPriceAsMana,
  [NotificationType.BID_RECEIVED]: formatPriceAsMana,
  [NotificationType.ROYALTIES_EARNED]: formatRoyaltiesCutAsMana
}

function loadTemplates(part: string) {
  return Object.values(NotificationType)
    .map((type) => fs.readFileSync(path.join(__dirname, `email-templates/${type}.${part}.handlebars`), 'utf8'))
    .reduce(
      (acc, template, index) => {
        acc[Object.values(NotificationType)[index]] = handlebars.compile(template, {})
        return acc
      },
      {} as Record<NotificationType, HandlebarsTemplateDelegate>
    )
}

export async function createRenderer(components: Pick<AppComponents, 'subscriptionService'>): Promise<IEmailRenderer> {
  const { subscriptionService } = components

  const contentTemplates = loadTemplates('content')
  const subjectTemplates = loadTemplates('subject')

  async function renderEmail(notification: NotificationRecord): Promise<Email> {
    const subscription = await subscriptionService.findSubscriptionForAddress(notification.address)
    const transformed = (transformers[notification.type] || nonTransformer)(notification)
    return {
      to: subscription?.email || '',
      subject: subjectTemplates[notification.type](transformed),
      content: contentTemplates[notification.type](transformed)
    }
  }

  return {
    renderEmail
  }
}
