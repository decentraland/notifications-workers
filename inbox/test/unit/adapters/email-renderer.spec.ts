import { createEmailRenderer, IEmailRenderer, InboxTemplates } from '../../../src/adapters/email-renderer'

describe('email rendering tests', () => {
  let renderer: IEmailRenderer

  type TestCaseData = {
    to: string
    context: any
  }

  const testCases: Record<InboxTemplates, TestCaseData> = {
    [InboxTemplates.VALIDATE_EMAIL]: {
      to: 'email@example.com',
      context: {}
    }
  }

  beforeAll(async () => {
    renderer = await createEmailRenderer()
  })

  const cases: [InboxTemplates, TestCaseData][] = Object.keys(testCases).map((type: InboxTemplates) => [
    type,
    testCases[type]
  ])

  test.each(cases)(`rendering %s`, async (type: InboxTemplates, data: TestCaseData) => {
    expect(await renderer.renderEmail(type, data.to, data.context)).toMatchSnapshot()
  })
})
