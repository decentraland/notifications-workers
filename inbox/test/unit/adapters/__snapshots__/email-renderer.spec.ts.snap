// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email rendering tests rendering validate-email 1`] = `
{
  "content": "<h1>Decentraland</h1>
<p>Hello!</p>
<p>You're receiving this email because you signed up for Decentraland email notifications.</p>
<p>You'll be able to customize exactly what kind of notifications you'd like to receive (e.g., Marketplace, DAO, Events,
    etc.), but you must first confirm your email.</p>

<p style="height: 41px; padding-top: 8px; padding-bottom: 8px;"><a
        href="" target="_blank"
        rel="noopener"
        style="
            background-color: #ff2d55;
            width: 365px;
            height: 41px;
            border-radius: 6px;
            color: #fff;
            padding: 12px 44px 12px 44px;
            text-decoration: none;
            font-family: system-ui, BlinkMacSystemFont, 'Segoe UI',
              Roboto, 'Helvetica Neue', Arial, sans-serif;
          "></a>
</p>

<p>Or you can copy and paste this URL into the browser: <span style="font-size: 13px;"></span>.
</p>
",
  "subject": "Confirm your email",
  "to": "email@example.com",
}
`;
